FFmpegInterop/H264SampleProvider.cpp

```diff
@@ -19,54 +19,136 @@
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_data[0] != 1);
 	}
 
-	// We have a complete frame
-	return hr;
-}
+	uint8_t AVCConfigParser::GetNaluLengthSize() const noexcept
+	{
+		return m_data[4] & 0x03 + 1;
+	}
 
-HRESULT H264SampleProvider::GetSPSAndPPSBuffer(DataWriter^ dataWriter)
-{
-	HRESULT hr = S_OK;
+	tuple<vector<uint8_t>, vector<uint32_t>> AVCConfigParser::GetSpsPpsData() const
+	{
+		vector<uint8_t> spsPpsData;
+		vector<uint32_t> spsPpsNaluLengths;
+		uint32_t pos{ 5 };
+
+		// Parse SPS NALUs
+		uint8_t spsCount{ static_cast<uint8_t>(m_data[pos++] & 0x1F) };
+		pos += ParseParameterSets(spsCount, m_data + pos, m_dataSize - pos, spsPpsData, spsPpsNaluLengths);
 
-	if (!m_pAvCodecCtx || (m_pAvCodecCtx->extradata == nullptr && m_pAvCodecCtx->extradata_size < 8))
+		// Parse PPS NALUs
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, pos >= m_dataSize);
+		uint8_t ppsCount{ m_data[pos++] };
+		ParseParameterSets(ppsCount, m_data + pos, m_dataSize - pos, spsPpsData, spsPpsNaluLengths);
+
+		return { spsPpsData, spsPpsNaluLengths };
+	}
+
+	uint32_t AVCConfigParser::ParseParameterSets(
+		_In_ uint8_t parameterSetCount,
+		_In_reads_(dataSize) const uint8_t* data,
+		_In_ uint32_t dataSize,
+		_Inout_ vector<uint8_t>& spsPpsData,
+		_Inout_ vector<uint32_t>& spsPpsNaluLengths) const
 	{
-		// The data isn't present
-		hr = E_FAIL;
+		// Reserve estimated space now to minimize reallocations
+		spsPpsNaluLengths.reserve(spsPpsNaluLengths.size() + parameterSetCount);
+
+		// Parse the parameter sets and convert the SPS/PPS data to Annex B format
+		uint32_t pos{ 0 };
+		for (uint8_t i{ 0 }; i < parameterSetCount; i++)
+		{
+			// Get the NALU length
+			uint32_t naluLength{ GetAVCNaluLength(data + pos, dataSize - pos, sizeof(uint16_t)) };
+			pos += sizeof(uint16_t);
+
+			// Write the NALU start code and SPS/PPS data to the buffer
+			spsPpsData.insert(spsPpsData.end(), begin(NALU_START_CODE), end(NALU_START_CODE));
+			spsPpsData.insert(spsPpsData.end(), data + pos, data + pos + naluLength);
+
+			spsPpsNaluLengths.push_back(sizeof(NALU_START_CODE) + naluLength);
+
+			pos += naluLength;
+		}
+
+		return pos; // Return the number of bytes read
 	}
-	else
+
+	AVCSequenceParameterSetParser::AVCSequenceParameterSetParser(_In_reads_(dataSize) const uint8_t* data, _In_ int dataSize)
 	{
-		// Write both SPS and PPS sequence as is from extradata
-		auto vSPSPPS = ref new Platform::Array<uint8_t>(m_pAvCodecCtx->extradata, m_pAvCodecCtx->extradata_size);
-		dataWriter->WriteBytes(vSPSPPS);
+		THROW_HR_IF_NULL(MF_E_INVALID_FILE_FORMAT, data);
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, dataSize < 3);
+
+		m_profile = data[1];
+
+		uint8_t profileCompatibility{ data[2] };
+		m_constraintSet1 = (profileCompatibility & (1 << 6)) != 0;
 	}
 
-	return hr;
-}
\ No newline at end of file
+	AVCPictureParamterSetParser::AVCPictureParamterSetParser(_In_reads_(dataSize) const uint8_t* data, _In_ int dataSize)
+	{
+		// TODO: Implement
```

Comment:
It seems like there's some missing work for this? (Also, TODO on MF_MT_VIDEO_H264_NO_FMOASO.) Put some information about what's planned here under "Expose more media types" in your description.

Reply:
H264 playback is working. The remaining work is just for MF_MT_VIDEO_H264_NO_FMOASO which is an H264 decoder optimization. From mfapi.h: "MF_MT_VIDEO_H264_NO_FMOASO set to non-zero (true) means external users/apps know that H.264 input video bitstream has no FMO/ASO enabled, that is, even if the bitstream has baseline profile and constraint_set1_flag equal to 0, the bitstream shall not have FMO/ASO then H.264 decoder uses DXVA decoding and doesn't fall back to software decoding it improves power consumption, memory usage, performance and user experiences (without unnecessary glitches on low end devices)."

Reply:
I added a bitstream reader for SPS/PPS parsing and implemented the MF_MT_VIDEO_H264_NO_FMOASO logic.